chainindex/reconcile.go

```diff
@@ -0,0 +1,211 @@
+		log.Infof("Marking tipsets as reverted from height %d", reconciliationEpoch)
+		result, err := tx.StmtContext(ctx, si.updateTipsetsToRevertedFromHeightStmt).ExecContext(ctx, int64(reconciliationEpoch))
+		if err != nil {
+			return xerrors.Errorf("failed to mark tipsets as reverted: %w", err)
+		}
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			return xerrors.Errorf("failed to get number of rows affected: %w", err)
+		}
+		log.Infof("Marked %d tipsets as reverted from height %d", rowsAffected, reconciliationEpoch)
+
+		// also need to mark events as reverted for the corresponding inclusion tipsets
+		if _, err = tx.StmtContext(ctx, si.updateEventsToRevertedFromHeightStmt).ExecContext(ctx, int64(reconciliationEpoch-1)); err != nil {
+			return xerrors.Errorf("failed to mark events as reverted: %w", err)
+		}
+
+		return si.applyMissingTipsets(ctx, tx, missingTipsets)
+	})
+}
+
+func (si *SqliteIndexer) backfillEmptyIndex(ctx context.Context, tx *sql.Tx, head *types.TipSet) error {
+	currTs := head
+	var missingTipsets []*types.TipSet
+
+	log.Infof("backfilling empty chain index from head height %d", head.Height())
+	var err error
+
+	for currTs != nil && len(missingTipsets) < si.maxReconcileTipsets {
+		missingTipsets = append(missingTipsets, currTs)
+		if currTs.Height() == 0 {
+			break
+		}
+
+		currTs, err = si.cs.GetTipSetFromKey(ctx, currTs.Parents())
+		if err != nil {
+			return xerrors.Errorf("failed to walk chain: %w", err)
+		}
+	}
+
+	return si.applyMissingTipsets(ctx, tx, missingTipsets)
+}
+
+func (si *SqliteIndexer) applyMissingTipsets(ctx context.Context, tx *sql.Tx, missingTipsets []*types.TipSet) error {
+	if len(missingTipsets) == 0 {
+		log.Info("No missing tipsets to index; index is all caught up with the chain")
+		return nil
+	}
+
+	log.Infof("Applying %d missing tipsets to Index; max missing tipset height %d; min missing tipset height %d", len(missingTipsets),
+		missingTipsets[0].Height(), missingTipsets[len(missingTipsets)-1].Height())
+	totalIndexed := 0
+
+	// apply all missing tipsets from the canonical chain to the current chain head
+	for i := 0; i < len(missingTipsets); i++ {
+		currTs := missingTipsets[i]
+		var parentTs *types.TipSet
+		var err error
+
+		if i < len(missingTipsets)-1 {
+			parentTs = missingTipsets[i+1]
+		} else if currTs.Height() > 0 {
+			parentTs, err = si.cs.GetTipSetFromKey(ctx, currTs.Parents())
+			if err != nil {
+				return xerrors.Errorf("failed to get parent tipset: %w", err)
+			}
+		} else if currTs.Height() == 0 {
+			if err := si.indexTipset(ctx, tx, currTs); err != nil {
+				log.Warnf("failed to index genesis tipset during reconciliation: %s", err)
+			} else {
+				totalIndexed++
+			}
+			break
+		}
+
+		if err := si.indexTipsetWithParentEvents(ctx, tx, parentTs, currTs); err != nil {
+			log.Warnf("failed to index tipset with parent events during reconciliation: %s", err)
```

Comment:
this again would be a good place to check for `NotFound` or some other signal that this is _normal_ and users shouldn't be worried about it

alternatively, make the message more verbose that this could be expected if the entire chain state isn't available, or something to soften it up

Reply:
@rvagg fixed this by returning an error from this function if we run into an error that is NOT an  `ipld. NotFound` and logging info if it is an `ipld. NotFound`. This works well.